### PR TITLE
fix(NODE-4999): Write Concern 0 Must Not Affect Read Operations

### PR DESCRIPTION
**Pull request by W-A-James**
### Description

Update operations that create cursors to set their writeConcern to undefined.

#### What is changing?

- `ChangeStream` now silently unsets the `writeConcern` option
-  `AggregateOperation` now silently unsets the `writeConcern` option
-  `FindOperation`  now silently unsets the `writeConcern` option
-  `ListIndexesOperation` now silently unsets the `writeConcern` option
-  `ListCollections` now silently unsets the `writeConcern` option

##### Is there new documentation needed for these changes?

No

#### What is the motivation for this change?

This fixes the bug where read operations that have a writeConcern of 0 fail when getMore is called.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/read-write-concern/write_concern.test.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
-'use strict';
-
 const { expect } = require('chai');
-const { LEGACY_HELLO_COMMAND } = require('../../../src/constants');
 
-const mock = require('../../tools/mongodb-mock/index');
 const { MongoClient } = require('../../../src');
+const { LEGACY_HELLO_COMMAND } = require('../../../src/constants');
```
Comment (W-A-James) [inline test/integration/read-write-concern/write_concern.test.js:4]: Should I go ahead and convert this file to a typescript file?

Comment (nbbeeken) [reply test/integration/read-write-concern/write_concern.test.js:4]: Sure!

File: test/integration/read-write-concern/write_concern.test.js [hunk 2]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
         });
     });
   });
+
+  describe('must not affect read operations', function () {
+    describe('when writeConcern = 0', function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:77]:
```suggestion
  context('when performing read operations', function () {
    context('when writeConcern = 0', function () {
```

Comment (durran) [reply test/integration/read-write-concern/write_concern.test.js:77]: Just to word and organise per our testing patterns.

File: test/integration/read-write-concern/write_concern.test.js [hunk 3]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+      describe('does not throw an error when getMore is called on cursor', function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:78]:
```suggestion
      describe('cursor creating operations with a getMore', function () {
```

File: test/integration/read-write-concern/write_concern.test.js [hunk 4]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+
+          await col.insertMany(docs);
+        });
+
+        afterEach(async function () {
+          await db.dropDatabase();
+          await client.close();
+        });
+
+        it('find', async function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:102]:
```suggestion
        it('succeeds on find', async function () {
```

File: test/integration/read-write-concern/write_concern.test.js [hunk 5]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+          const findResult = col.find({}, { batchSize: 2 });
+          const err = await findResult.toArray().catch(e => e);
+
+          expect(err).to.not.be.instanceOf(Error);
+        });
+
+        it('listCollections', async function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:109]:
```suggestion
        it('succeeds on listCollections', async function () {
```

File: test/integration/read-write-concern/write_concern.test.js [hunk 6]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+          }
+
+          const cols = db.listCollections({}, { batchSize: 2 });
+
+          const err = await cols.toArray().catch(e => e);
+
+          expect(err).to.not.be.instanceOf(Error);
+        });
+
+        it('aggregate', async function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:126]:
```suggestion
        it('succeeds on aggregate', async function () {
```

File: test/integration/read-write-concern/write_concern.test.js [hunk 7]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+          const aggResult = col.aggregate([{ $match: { a: { $gte: 0 } } }], { batchSize: 2 });
+          const err = await aggResult.toArray().catch(e => e);
+
+          expect(err).to.not.be.instanceOf(Error);
+        });
+
+        it('listIndexes', async function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:133]:
```suggestion
        it('succeeds on listIndexes', async function () {
```

File: test/integration/read-write-concern/write_concern.test.js [hunk 8]
```diff
@@ -72,4 +72,91 @@ describe('Write Concern', function () {
...
+          await col.createIndex({ b: -1 });
+          await col.createIndex({ a: 1, b: -1 });
+
+          const listIndexesResult = col.listIndexes({ batchSize: 2 });
+          const err = await listIndexesResult.toArray().catch(e => e);
+
+          expect(err).to.not.be.instanceOf(Error);
+        });
+
+        it('changeStream', async function () {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.js:144]:
```suggestion
        it('succeeds on changeStream', async function () {
```

**Review by durran — CHANGES_REQUESTED** [review]

File: test/integration/read-write-concern/write_concern.test.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+        it('succeeds on changeStream', async function () {
+          if (this.configuration.options.replicaSet) {
+            const changeStream = col.watch(undefined, { batchSize: 2 });
+
+            setTimeout(() => {
+              col.updateMany({}, [{ $addFields: { A: 1 } }]);
+            }, 1000);
+
+            const err = await changeStream.next().catch(e => e);
```
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/read-write-concern/write_concern.test.ts:153]: I'm pulling down and having a look.

File: test/integration/read-write-concern/write_concern.test.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+            expect(err).to.not.be.instanceOf(Error);
+          } else {
+            this.skip();
+          }
+        });
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.ts:159]:
```suggestion
        it('succeeds on changeStream', {
          metadata: { requires: { topology: 'replicaset' } },
          async test() {
            const changeStream = col.watch([], { batchSize: 2 });

            setTimeout(() => {
              col.updateMany({}, [{ $addFields: { A: 1 } }]);
            }, 1000);

            const err = await changeStream.next().catch(e => e);

            expect(err).to.not.be.instanceOf(Error);
          }
        });
```

**Review by durran — CHANGES_REQUESTED** [review]

File: test/integration/read-write-concern/write_concern.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/read-write-concern/write_concern.test.ts:153]: I've added a suggestion now that should work.

[the remark this answers is not in the document]
Comment (durran) [reply test/integration/read-write-concern/write_concern.test.ts:153]: Sounds like a good idea.

File: src/operations/aggregate.ts [hunk 11]
```diff
@@ -65,6 +65,8 @@ export class AggregateOperation<T = Document> extends CommandOperation<T> {
 
     if (this.hasWriteStage) {
       this.trySecondaryWrite = true;
+    } else {
+      this.options.writeConcern = undefined;
```
[the remark this answers is not in the document]
Comment (durran) [reply src/operations/aggregate.ts]: Sounds like a good idea.

File: test/integration/read-write-concern/write_concern.test.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+          await col.createIndex({ b: -1 });
+          await col.createIndex({ a: 1, b: -1 });
+
+          const listIndexesResult = col.listIndexes({ batchSize: 2 });
+          const err = await listIndexesResult.toArray().catch(e => e);
+
+          expect(err).to.not.be.instanceOf(Error);
+        });
+
+        it('succeeds on changeStream', {
```
Comment (durran) [inline test/integration/read-write-concern/write_concern.test.ts:145]: Ok so it's now limited to 3.6 and 4.0 of the server. Wondering why that's the case...

Comment (W-A-James) [reply test/integration/read-write-concern/write_concern.test.ts:145]: Testing locally now

Comment (W-A-James) [reply test/integration/read-write-concern/write_concern.test.ts:145]: Stepping through in the debugger, we're getting stuck on `src/change_stream.ts:692` at the `this.cursor.next`. Looking into what's going wrong with it.

Comment (durran) [reply test/integration/read-write-concern/write_concern.test.ts:145]: Maybe it's just not getting changes and version is a red herring. Maybe try refactoring it to use the events module and awaiting the next there like we do in the other tests? See: https://github.com/mongodb/node-mongodb-native/blob/main/test/integration/change-streams/change_stream.test.ts#L177

**Review by durran — DISMISSED** [review]
LGTM - confirmed CI failures are unrelated (flaky FLE prose test)

File: test/integration/read-write-concern/write_concern.test.ts [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    let server;
+    before(() => {
+      return mock.createServer().then(s => {
+        server = s;
+      });
+    });
+
+    after(() => mock.cleanup());
+
+    // TODO: NODE-3816
```
Comment (nbbeeken) [inline test/integration/read-write-concern/write_concern.test.ts:50]:
```suggestion
    // TODO(NODE-3816): the mock server response is looking for writeConcern on all messages, but endSessions doesn't have it
```

Just cloning some context from the ticket + aligning the todo format with how we normally have the comment.

**Review by nbbeeken — COMMENTED** [review]
Minor nit about a prior todo

**Review by durran — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**